kong/init.lua

```diff
@@ -269,10 +269,7 @@ local function execute_plugins_iterator(plugins_iterator, phase, ctx)
       local cok, cerr = coroutine.resume(co, plugin.handler, configuration)
       if not cok then
         kong.log.err(cerr)
-        ctx.delayed_response = {
-          status_code = 500,
-          content = { message  = "An unexpected error occurred" },
-        }
+        kong.response.error(500)
```

Comment:
This doesn't look right. We need to collect the plugins even in case of an error.

Reply:
What do you mean by `collect the plugins`

Reply:
@Elyahou, in access phase we need to run through the whole plugin iterator. You cannot exit it while looping.

Reply:
That's not correct and is not the case already before my changes...

You can see that on line 267 in this file there is a condition before running each plugin `elseif not ctx.delayed_response then`, and this is precisely the variable that is set when a plugin execution error occur.

Reply:
@Elyahou, I think I am correct. Ypi added `kong.response.error(500)` that will call `ngx.exit` and jump to `header_filter` phase. You are effectively exiting the loop that collects the plugins:

```lua
for plugin, configuration in plugins_iterator:iterate(phase, ctx) do
```

That loop need to be run full, even when we don't run the plugins themselves. Is this clear enough?

Reply:
@bungle inside this loop, there is a condition to run plugin when `delay_response ` is true, meaning we are in the access phase, this condition is at line 267:

```
elseif not ctx.delayed_response then
```

Meaning that is ` ctx.delayed_response` is filled, the next plugin will not run.  But `ctx.delayed_response` is precisely the variable that is set when there is an error in a plugin before my changes, meaning that even before my changes, a plugin error cause the iterator to not run the next plugins...

Is this clear ?

Reply:
Also, when we are in access phase, the `Response.error` function will not exit directly but will add the response to `ctx.delayed_response`: https://github.com/Kong/kong/blob/7e6f8f38a8ff9253ccfe48e5defd17cf15d9aa75/kong/pdk/response.lua#L1028-L1038

Reply:
@Elyahou, again, it will end here:
https://github.com/Kong/kong/blob/master/kong/pdk/response.lua#L1055

This will now `yield` the main thread. And this is not what we want. We want the whole iteration to happen as `plugins_iterator:iterate(phase, ctx)` is the thing that collects the plugins. See here we create sub-thread:
https://github.com/Kong/kong/blob/master/kong/init.lua#L278-L279

This is what could be `yielded`, and that is what happens if plugin errors on `access`.

Reply:
Thanks @bungle for the details, it's weird because I can see that even with these changes, the iterator still go through every plugin on access phase... Even after a plugin error. IT seems like the coroutine yield so not stop it...

Anyway, I can change it to call `Response.error` inside coroutine to avoid this issue, something like:

```lua
local function access_phase_runner(access_function, handler, configuration)
  return function(handler, conf)
    local ok, err = pcall(access_function, handler, conf)

    if not ok then
      kong.log.err(err)
      kong.response.error(500)
    end
  end
end

local function execute_plugins_iterator(plugins_iterator, phase, ctx)
...
for plugin, configuration in plugins_iterator:iterate(phase, ctx) do
...
    if not delay_response then
      plugin.handler[phase](plugin.handler, configuration)

    elseif not ctx.delayed_response then
      local co = coroutine.create(access_phase_runner(plugin.handler.access))
      coroutine.resume(co, plugin.handler, configuration)
    end
    ...
  end
...
end
```

What do you think ?

Reply:
@Elyahou, yes it could be that `yielding` main thread will just make worker to jump to other thread and resume later. This seems to be confirming that: https://github.com/openresty/lua-nginx-module#ngxthreadspawn. Yielding main thread gives CPU to other thread, so that is a slightly different than what it originally was, though the end result is the same.

Reply:
Careful about returning closures on the proxy's hot code path though - they have a noticeable effect on memory usage (GC pressure) and overall throughput.

Reply:
See this commit for background info on this part of the runloop (delayed responses): https://github.com/Kong/kong/commit/e709c953933e5ca6f6e9766fde15d8fe98a80721

Reply:
Thank you @thibaultcha for these references. This coroutine yielding cause me to think that the `Response.error` function should not be called from outside a plugin at all... But still, I think errors formatting should be centralised... So maybe, we should extract the formating logic outside to some `utils` and use it here and also in the `Response.error` function. What do you think ?